frontend/src/Frontend/VersionedStore/V2.hs

```diff
@@ -0,0 +1,216 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE TemplateHaskell #-}
+module Frontend.VersionedStore.V2 where
+
+import Data.Aeson
+import Data.Aeson.GADT.TH
+import Data.Constraint (Dict(Dict))
+import Data.Constraint.Extras
+import Data.Dependent.Map (DMap)
+import Data.Dependent.Sum (DSum(..))
+import qualified Data.Dependent.Map as DMap
+import Data.Function (on)
+import Data.Functor ((<&>))
+import Data.Functor.Identity (Identity(Identity), runIdentity)
+import qualified Data.IntMap as IntMap
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Data.Maybe (catMaybes)
+import Data.Text (Text)
+import Text.Printf (printf)
+
+import Common.Foundation
+import Common.Wallet
+import Common.Network (NetworkName, NodeRef, parseNodeRef)
+import Common.OAuth (OAuthProvider(..))
+import Common.GistStore (GistMeta)
+
+import Frontend.VersionedStore.TH
+import qualified Frontend.VersionedStore.V0 as V0
+import qualified Frontend.VersionedStore.V1 as V1
+import qualified Frontend.VersionedStore.V0.Wallet as V0
+import Frontend.VersionedStore.MigrationUtils
+import Frontend.Crypto.Class
+
+-- WARNING: Upstream deps. Check this when we bump pact and obelisk!
+-- May be worth storing this in upstream independent datatypes.
+import Pact.Types.ChainId (ChainId(ChainId))
+import Pact.Types.ChainMeta (PublicMeta (..))
+import Obelisk.OAuth.Common (AccessToken, OAuthState)
+
+data StoreFrontend key a where
+  StoreFrontend_Wallet_Keys :: StoreFrontend key (KeyStorage key)
+  StoreFrontend_Wallet_Accounts :: StoreFrontend key AccountStorage
+
+  StoreFrontend_Network_PublicMeta :: StoreFrontend key PublicMeta
+  StoreFrontend_Network_Networks :: StoreFrontend key (Map NetworkName [NodeRef])
+  StoreFrontend_Network_SelectedNetwork :: StoreFrontend key NetworkName
+
+  StoreFrontend_OAuth_Tokens :: StoreFrontend key (Map OAuthProvider AccessToken)
+  StoreFrontend_OAuth_State :: OAuthProvider -> StoreFrontend key OAuthState
+
+  StoreFrontend_Gist_GistRequested :: StoreFrontend key (GistMeta, Text)
+
+  StoreFrontend_ModuleExplorer_SessionFile :: StoreFrontend key Text
+
+deriving instance Show (StoreFrontend key a)
+
+upgradeFromV0 :: (Monad m, HasCrypto key m) => DMap (V0.StoreFrontend key) Identity -> m (DMap (StoreFrontend key) Identity)
+upgradeFromV0 v0 = do
+  (newKeysList, newAccountStorage) <- foldMapM splitOldKey oldKeysList
```

Comment:
`splitOldKey` really concerns me, for the web version. Its relying on the keys' index in a list, and then using that index to "regenerate" them instead of saving /copying them somewhere. This code is so easy to accidentally copy over, for, say the next migration. We should either rewrite it or at the very least use CPP to make it a no-op on ghcjs. Not quite sure its in the scope of this task, but it could so easily go really bad really quickly if it isn't taken care of imo